Attempts to repair more of the closing tags.

aerych opened the pull request:
Closes #20 
Attempts to repair more than just a truncated closing tag.

aerych commented:
cc @astralbodies Would you be up for a very vicious review of this change?

aerych commented:
I just realized @astralbodies is going to be out of pocket for a bit.  @sendhil or @diegoreymendez would either of you mind taking a look?

diegoreymendez commented:
@aerych - Will do!

diegoreymendez commented on WPXMLRPC/WPXMLRPCDataCleaner.m:
```diff
@@ -82,18 +84,18 @@ @implementation WPXMLRPCDataCleaner (CleaningSteps)
+    char *outptr = outbuf;
+
+    if (iconv(cd, &inbuf, &inbytesleft, &outptr, &outbytesleft) == (size_t)-1) {
+        // Failed iconv, possible errors to encounter in `errno`:
```
Just double checking since I don't see any obvious difference here... you just changed the indentation, correct?

aerych replied:
An unintentional change.  I think Xcode took care of the formatting due to some copy/pasting.  Since its switching tabs to spaces its actually more inline with our coding standards.  Call it a happy accident. :)

diegoreymendez replied:
All good... was a bit worried I may be missing something else.

diegoreymendez commented on WPXMLRPC/WPXMLRPCDataCleaner.m:
```diff
@@ -191,30 +193,54 @@ - (NSString *)cleanWithTidyIfPresent:(NSString *)str {
+
+    // If the length of the content cleaned from before the preamble is greater
+    // than the length of the closing tags, then the xml is too damaged to repair.
+    if (length > closingTags) {
```
Should probably be:

`if (length > [closingTags length]) {`

aerych replied:
DoH! Yes. Yes it should.

diegoreymendez commented on WPXMLRPC/WPXMLRPCDataCleaner.m:
```diff
@@ -191,30 +193,54 @@ - (NSString *)cleanWithTidyIfPresent:(NSString *)str {
 
+    // Find the ending stub
+    NSString *stub = [str substringFromIndex:range.location];
+    range = [closingTags rangeOfString:stub];
```
Not sure if this is possible, so please bear with me..

Is it possible str could end in `</param><`?  If so, then the ending stub would be `<`, and the code would duplicate the last `</param>` tag ending the variable str with `</param></param></params></methodResponse>`.

If this was possible, then it could be a better idea to look at the end of `str` for any partial match of `closingTags` directly.

aerych replied:
Oh good catch! Yes, it does indeed look like that is possible. Will fix.

diegoreymendez commented:
@aerych - First run is finished... anything you need let me know!

aerych commented:
Thanks @diegoreymendez.  I've updated with changes, and added some unit tests.  Ready for a second look if you're available.  Otherwise pinging @sendhil :)

diegoreymendez commented on WPXMLRPCTest/Tests/WPXMLRPCDataCleanerTest.m:
```diff
@@ -11,20 +11,61 @@ - (void)testClosingTagRepaired
+    STAssertEqualObjects(completeResponse, cleanedString, @"A good reponse should just be returned intact.");
 
-    completeResponse = @"<methodResponse><params><param><value><string>Hello!</string></value></param></params></methodResponse>";
+    // Test a whole response but some info was cleaned from before the preamble
```
First off, very nice work with the amount of unit tests provided.

However, it seems to me like it could be a good idea to break these up into different tests.  It would make the code slightly more readable and the tests easier to maintain.

aerych replied:
Fair point. I considered it but it felt like more boiler plate just to test the different test cases for a single method. Simple to add more methods if it feels easier to maintain.

diegoreymendez commented on WPXMLRPC/WPXMLRPCDataCleaner.m:
```diff
@@ -191,30 +193,77 @@ - (NSString *)cleanWithTidyIfPresent:(NSString *)str {
+- (NSString *)cleanClosingTagIfNeeded:(NSString *)str lengthOfCharactersPrecedingPreamble:(NSInteger)length
 {
+    // Clean responses, but not requests (for now).
+    if ([str rangeOfString:@"methodResponse"].location == NSNotFound) {
```
This is a purely style comment.  I'm being a bit nitpicky, so feel free to disagree.

I couldn't find any obvious logic mistakes in this code.  However one thing that I did notice when reading this method is that following the logic is a bit difficult.  I feel this is backed up a bit by the amount of comments in the code.

There are two suggestions I can make, to make this method slightly easier to follow:

1. Add some extra `BOOL`s to make the conditions being checked more obvious to the reader.  For example instead of:

```objective-c
if ([str rangeOfString:@"methodResponse"].location == NSNotFound) {
```

Use:

```objective-c
BOOL strContainsAResponse = ([str rangeOfString:@"methodResponse"].location != NSNotFound);

if (!strContainsAResponse) {
```

2. Break this method up into sub methods so that instead of having these lines inline...

```objective-c
NSString *closingTags;
if ([str rangeOfString:@"<params>"].location == NSNotFound) {
    closingTags = @"</value></fault></methodResponse>";
} else {
    closingTags = @"</param></params></methodResponse>";
}
```

...we have...

```objective-c
NSString *closingTags = [self closingTagsForRequest:str];
```

diegoreymendez commented:
Eric - Second review is complete.  If you happen to agree on my comments, but prefer to leave that for later, feel free to open a different issue and ping me back so I approve this PR.

Also if you disagree just ping me back and I'll approve your PR since I didn't find any problems with it.

aerych commented:
@diegoreymendez Updated for readability.  Round three? :)

diegoreymendez commented:
Aight, thanks @aerych let's :shipit: !
